semgrep-core/src/core/Rule.ml

```diff
@@ -226,20 +216,34 @@ and paths = {
 }
 [@@deriving show]
 
+type search_rule = [ `Search of common * pformula ] [@@deriving show]
+type taint_rule = [ `Taint of common * taint_spec ] [@@deriving show]
+type rule = [ search_rule | taint_rule ] [@@deriving show]
+
 (* alias *)
 type t = rule [@@deriving show]
 type rules = rule list [@@deriving show]
 
+let id (`Search (c, _) | `Taint (c, _)) = c.id
```

Comment:
Having to write all these helpers and keep them updated is pretty annoying, wonder if there's a ppx deriving extension that will do this?

Reply:
I don't think there would be, and if we have some fields only shared for a couple modes then we'll probably have to do something like this anyway unfortunately. I don't think it'll be too burdensome though.

Reply:
It may be worth considering having a `common : rule -> common` function, and then write all the functions below in terms of that, e.g. `let message rule = (common rule).message`.

Reply:
My only concern there would be wrt how "common" some of the data search and taint share would be to other modes (e.g., `severity`, `fix`)

Reply:
We could also make a polymorphic rule_
type 'mode rule_ = {
   id: ...
   message:
   mode: 'mode;
 }
and then define the mode = `Tainting of ... | `Search of ...
and then you can type rule = mode rule_
type tainting_rule = [`Tainting of ...] rule_
?

Reply:
I agree with @aryx here it's worth trying this. I believe subtyping also applies when using polymorphic variants as type parameters.

Reply:
> I believe subtyping also applies when using polymorphic variants as type parameters.

I couldn't find docs which mention variance explicitly, but a couple examples in the repl seemed to be fine, so I'm thinking the same.

Going to implement it this way and see.